gax/src/main/java/com/google/api/gax/rpc/ReframingResponseObserver.java

```diff
@@ -0,0 +1,364 @@
+ * ApiCallContext context) {
+ *      Reframer<Chunk, FullResponse> myReframer = new Reframer<>();
+ *      innerCallable.call(request, new ReframingResponseObserver(myReframer, outerObserver),
+ * context);
+ *    }
+ *  }
+ * }</pre>
+ */
+@BetaApi("The surface for streaming is not stable yet and may change in the future.")
+public class ReframingResponseObserver<InnerT, OuterT> implements ResponseObserver<InnerT> {
+  private final Object lock = new Object();
+
+  @GuardedBy("lock")
+  private final ResponseObserver<OuterT> outerResponseObserver;
+
+  @GuardedBy("lock")
+  private final Reframer<OuterT, InnerT> reframer;
+
+  private StreamController innerController;
+  private boolean hasStarted;
+  private boolean autoFlowControl = true;
+
+  @GuardedBy("lock")
+  private int numPending;
+
+  /** Delivery mutual exclusion: only one thread can be delivering at a time */
+  @GuardedBy("lock")
+  private boolean inDelivery;
+  /** When another thread is delivering, signal it to take an extra run to pick up changes */
+  @GuardedBy("lock")
+  private boolean missed;
+
+  // Request mutex: only one upstream request can be active at a time
+  @GuardedBy("lock")
+  private boolean awaitingInner;
+  // By default all errors should be deferred until buffers have been exhausted
+  // However, when a stream is cancelled, the errors should be immediate
+  @GuardedBy("lock")
+  private boolean deferError = true;
+  // Deferred delivery stop: the upstream is exhausted, request downstream notification when internal buffers are exhausted
+  @GuardedBy("lock")
+  private boolean closeOnDone;
+  // When closing signal error.
+  @GuardedBy("lock")
+  private Throwable error;
+  // This stream has been closed, prevent further processing.
+  @GuardedBy("lock")
+  private boolean closed;
+
+  public ReframingResponseObserver(
+      ResponseObserver<OuterT> observer, Reframer<OuterT, InnerT> reframer) {
+    this.outerResponseObserver = observer;
+    this.reframer = reframer;
+  }
+
+  /**
+   * Callback that will be notified when the inner/upstream callable starts. This will in turn
+   * notify the outer/downstreamObserver of stream start. Regardless of the
+   * outer/downstreamObserver, the upstream controller will be put into manual flow control.
+   *
+   * @param controller The controller for the upstream stream.
+   */
+  @Override
+  public void onStart(StreamController controller) {
+    innerController = controller;
+
+    outerResponseObserver.onStart(
+        new StreamController() {
+          @Override
+          public void disableAutoInboundFlowControl() {
+            Preconditions.checkState(
+                !hasStarted, "Can't disable automatic flow control once the stream has started");
+            autoFlowControl = false;
+            numPending = 0;
+          }
+
+          @Override
+          public void request(int count) {
+            ReframingResponseObserver.this.onRequest(count);
```

Comment:
Add a note for later code readers: `innerController.request(int)` is indirectly invoked.